### PR TITLE
Implement JsonContext and JsonTypeInfo Overloads for Binary Data

### LINKED ISSUE #54979: Add JsonTypeInfo overloads to System.Memory.Data
## Background and Motivation

In 6.0, we have added APIs to System.Text.Json and System.Net.Http.Json that take a JsonTypeInfo or JsonSerializerContext. See https://github.com/dotnet/runtime/issues/1568 for the reasons and advantages of these overloads.

We have very thin wrapper APIs in System.Memory.Data around JsonSerializer, however these APIs weren't updated for JsonTypeInfo overloads. We should add them to get the same advantages as listed above.

## Proposed API

```diff
namespace System
{
    public partial class BinaryData
    {
        // exisitng APIs
        public BinaryData(object? jsonSerializable, System.Text.Json.JsonSerializerOptions? options = null, System.Type? type = null) { }
        public static System.BinaryData FromObjectAsJson<T>(T jsonSerializable, System.Text.Json.JsonSerializerOptions? options = null) { throw null; }
        public T? ToObjectFromJson<T>(System.Text.Json.JsonSerializerOptions? options = null) { throw null; }

        // new proposed APIs
+       public BinaryData(object? jsonSerializable, System.Text.Json.Serialization.JsonSerializerContext context, System.Type? type = null) { }
+       public static System.BinaryData FromObjectAsJson<T>(T jsonSerializable, System.Text.Json.Serialization.Metadata.JsonTypeInfo<T> jsonTypeInfo) { throw null; }
+       public T? ToObjectFromJson<T>(System.Text.Json.Serialization.Metadata.JsonTypeInfo<T> jsonTypeInfo) { throw null; }
    }
}
```

## Usage Examples

``` C#
var myObject = new MyType();

var binaryData = new BinaryData(myObject, JsonContext.Default);
// --- or ---
var binaryData = BinaryData.FromObjectAsJson(myObject, JsonContext.Default.MyType);


var myObject2 = binaryData.ToObjectFromJson(JsonContext.Default.MyType);
```

## Alternative Designs

N/A

## Risks

None

cc @layomia @JoshLove-msft @KrzysztofCwalina

### PR DESCRIPTION
**Pull request by Tarun047**
Fixes #54979

- [x] Implemented Constructor Overload for Binary Data for working with ```JsonSerializerContext```
``` public BinaryData(object? jsonSerializable, JsonSerializerContext context, System.Type? type) ```
- [x] Implemented static Overload for ```FromObjectAsJson<T>``` for working with ```JsonTypeInfo<T>```
```public static System.BinaryData FromObjectAsJson<T>(T jsonSerializable, JsonTypeInfo<T> jsonTypeInfo) ```
- [x] Implemented static Overload for ```ToObjectFromJson<T>``` for working with  ```JsonTypeInfo<T>```
```public T? ToObjectFromJson<T>(JsonTypeInfo<T> jsonTypeInfo) ```
- [x] Implemented unit tests for above changes

### CODE REVIEW

File: src/libraries/System.Memory.Data/tests/BinaryDataTests.cs [hunk 1]
```diff
@@ -591,14 +598,30 @@ public void ToStringReturnEmptyStringWhenBinaryDataEmpty()
...
             public int B { get; set; }
             public bool C { get; set; }
             public object D { get; set; }
         }
 
+        internal class MismatchedTestModel 
+        {
+            public int A { get; set; }
+        }
+
```
Comment (adamsitnik) [inline src/libraries/System.Memory.Data/tests/BinaryDataTests.cs:613]: minor nit: extra empty line

```suggestion
```

File: src/libraries/System.Memory.Data/src/System/BinaryData.cs [hunk 2]
```diff
@@ -55,6 +57,23 @@ public BinaryData(object? jsonSerializable, JsonSerializerOptions? options = def
             _bytes = JsonSerializer.SerializeToUtf8Bytes(jsonSerializable, type, options);
         }
 
+         /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by serializing the provided object to JSON
```
Comment (adamsitnik) [inline src/libraries/System.Memory.Data/src/System/BinaryData.cs:61]: nit: extra space

```suggestion
        /// <summary>
        /// Creates a <see cref="BinaryData"/> instance by serializing the provided object to JSON
```

File: src/libraries/System.Memory.Data/src/System/BinaryData.cs [hunk 3]
```diff
@@ -189,6 +208,21 @@ public static BinaryData FromObjectAsJson<T>(T jsonSerializable, JsonSerializerO
...
+        /// <typeparam name="T">The type to use when serializing the data.</typeparam>
+        /// <param name="jsonSerializable">The data to use.</param>
+        /// <param name="jsonTypeInfo">The <see cref="JsonTypeInfo"/> to use when serializing to JSON.</param>
+        /// <returns>A value representing the UTF-8 encoding of the JSON representation of <paramref name="jsonSerializable" />.</returns>
+        [RequiresUnreferencedCode(JsonSerializerRequiresUnreferencedCode)]
+        public static BinaryData FromObjectAsJson<T>(T jsonSerializable, JsonTypeInfo<T> jsonTypeInfo)
+        {
+            byte[] buffer = JsonSerializer.SerializeToUtf8Bytes(jsonSerializable, jsonTypeInfo);
+            return new BinaryData(buffer);
+        }
```
Comment (adamsitnik) [inline src/libraries/System.Memory.Data/src/System/BinaryData.cs:224]: subjective (feel free to ignore): we could use expression syntax here

```suggestion
            => new BinaryData(JsonSerializer.SerializeToUtf8Bytes(jsonSerializable, jsonTypeInfo));
```

File: src/libraries/System.Memory.Data/src/System/BinaryData.cs [hunk 4]
```diff
@@ -249,6 +283,20 @@ public override unsafe string ToString()
...
+        /// </summary>
+        /// <typeparam name="T">The type that the data should be
+        /// converted to.</typeparam>
+        /// <param name="jsonTypeInfo">The <see cref="JsonTypeInfo"/> to use when serializing to JSON.</param>
+        /// <returns>The data converted to the specified type.</returns>
+        [RequiresUnreferencedCode(JsonSerializerRequiresUnreferencedCode)]
+        public T? ToObjectFromJson<T>(JsonTypeInfo<T> jsonTypeInfo)
+        {
+            return JsonSerializer.Deserialize<T>(_bytes.Span, jsonTypeInfo);
+        }
```
Comment (adamsitnik) [inline src/libraries/System.Memory.Data/src/System/BinaryData.cs:298]: same as above

```suggestion
            => JsonSerializer.Deserialize<T>(_bytes.Span, jsonTypeInfo);
```

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @Tarun047 

before I hit the merge button, please fix the whitespaces issues and let me know what do you think about my suggestions

File: src/libraries/System.Memory.Data/tests/BinaryDataTests.cs (hunk 1 above)
Comment (Tarun047) [reply src/libraries/System.Memory.Data/tests/BinaryDataTests.cs:613]: Done

File: src/libraries/System.Memory.Data/src/System/BinaryData.cs (hunk 2 above)
Comment (Tarun047) [reply src/libraries/System.Memory.Data/src/System/BinaryData.cs]: Fixed

File: src/libraries/System.Memory.Data/src/System/BinaryData.cs (hunk 4 above)
Comment (Tarun047) [reply src/libraries/System.Memory.Data/src/System/BinaryData.cs]: Done

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**